Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm

```diff
@@ -0,0 +1,292 @@
+#include <WebCore/GraphicsLayer.h>
+#include <WebCore/GraphicsLayerClient.h>
+#include <WebCore/IntPoint.h>
+#include <WebCore/Page.h>
+#include <pal/spi/cocoa/DataDetectorsCoreSPI.h>
+#include <pal/spi/mac/DataDetectorsSPI.h>
+#include <wtf/Algorithms.h>
+#include <wtf/OptionSet.h>
+#include <wtf/RefPtr.h>
+#include <wtf/RetainPtr.h>
+#include <wtf/Vector.h>
+#include <wtf/cocoa/VectorCocoa.h>
+#include <wtf/text/WTFString.h>
+
+#include <pal/mac/DataDetectorsSoftLink.h>
+
+namespace WebKit {
+
+using namespace WebCore;
+
+PDFDataDetectorOverlayController::PDFDataDetectorOverlayController(UnifiedPDFPlugin& plugin)
+    : m_plugin(plugin)
+{
+}
+
+RefPtr<UnifiedPDFPlugin> PDFDataDetectorOverlayController::protectedPlugin() const
+{
+    return m_plugin.get();
+}
+
+PageOverlay& PDFDataDetectorOverlayController::installOverlayIfNeeded()
+{
+    if (m_overlay)
+        return *m_overlay;
+
+    m_overlay = PageOverlay::create(*this, PageOverlay::OverlayType::Document, PageOverlay::AlwaysTileOverlayLayer::Yes);
+    protectedPlugin()->installDataDetectorOverlay(*m_overlay);
+
+    return *m_overlay;
+}
+
+void PDFDataDetectorOverlayController::uninstallOverlay()
+{
+    if (!m_overlay)
+        return;
+
+    RefPtr plugin = protectedPlugin();
+    if (!plugin)
+        return;
+
+    plugin->uninstallDataDetectorOverlay(*std::exchange(m_overlay, nullptr));
+}
+
+void PDFDataDetectorOverlayController::teardown()
+{
+    uninstallOverlay();
+}
+
+static RetainPtr<DDHighlightRef> createPlatformDataDetectorHighlight(Vector<FloatRect>&& highlightBounds, FloatRect&& visibleContentRect)
+{
+    Vector highlightBoundsCG = highlightBounds.map([](const FloatRect& bound) -> CGRect {
+        return bound;
+    });
+
+    DDHighlightStyle style = static_cast<DDHighlightStyle>(DDHighlightStyleBubbleStandard) | static_cast<DDHighlightStyle>(DDHighlightStyleStandardIconArrow);
+    BOOL drawButton = YES;
+    NSWritingDirection writingDirection = NSWritingDirectionNatural;
+    BOOL endsWithEOL = NO;
+    BOOL drawFlipped = YES;
+    float targetSurfaceBackingScaleFactor = 0;
+
+    return adoptCF(PAL::softLink_DataDetectors_DDHighlightCreateWithRectsInVisibleRectWithStyleScaleAndDirection(nullptr, highlightBoundsCG.data(), highlightBounds.size(), visibleContentRect, style, drawButton, writingDirection, endsWithEOL, drawFlipped, targetSurfaceBackingScaleFactor));
+}
+
+RetainPtr<DDHighlightRef> PDFDataDetectorOverlayController::createPlatformDataDetectorHighlight(PDFDataDetectorItem& dataDetectorItem) const
+{
+    RefPtr plugin = protectedPlugin();
+    if (!plugin)
+        return { };
+
+    auto selectionBoundsInPluginSpace = plugin->rectForSelectionInPluginSpace(dataDetectorItem.selection().get());
+    auto visibleContentRectInPluginSpace = plugin->visibleContentRect();
+
+    return ::WebKit::createPlatformDataDetectorHighlight(Vector<FloatRect>::from(WTFMove(selectionBoundsInPluginSpace)), WTFMove(visibleContentRectInPluginSpace));
```

Comment:
The `::WebKit` is necessary?

Reply:
Yes, only to disambiguate between the free function and the class method. I could also just rename the free function `createPlatformDataDetectorHighlightImpl` or something.

Reply:
Keeping it as-is for now.